utils/firestore.js

```diff
@@ -2553,21 +2553,83 @@ const addPrintAddressesParticipants = async (data) => {
+                            .select('Connect_ID', 
+                                `${healthCareProvider}`, 
+                                `${collectionDetails}`, 
+                                `${mouthwashSurveyCompletionStatus}`)
+                            .get();
+        
+        if (snapshot.size > 0) {
+            for (const docs of snapshot.docs) {
+                const participantData = docs.data();
+                const participantConnectID = participantData['Connect_ID'];
+                
+                const customParticipantObj = {
+                    "Connect_ID": participantConnectID,
+                    [healthCareProvider]: participantData[healthCareProvider],
+                    [shippedDateTime]: participantData[collectionDetails]?.[baseline]?.[bioKitMouthwash]?.[shippedDateTime] || '',
+                    [mouthwashSurveyCompletionStatus]: participantData[mouthwashSurveyCompletionStatus],
+                };
+                
+                try { 
+                    const { supplyKitId, supplyKitTrackingNum, returnKitId, collectionCardId, returnKitTrackingNum } = fieldMapping;
+                    const kitAssemblySnapshot = await db.collection("kitAssembly")
+                                                    .where('Connect_ID', '==', participantConnectID)
+                                                    .select(`${supplyKitId}`,
+                                                        `${supplyKitTrackingNum}`, 
+                                                        `${returnKitTrackingNum}`,
+                                                        `${returnKitId}`, 
+                                                        `${collectionCardId}`)
+                                                    .get();
+
+                    if (kitAssemblySnapshot.size > 0) { 
+                        const kitAssemblyData = kitAssemblySnapshot.docs[0].data();
+
+                        customParticipantObj[supplyKitId] = kitAssemblyData[supplyKitId];
+                        customParticipantObj[supplyKitTrackingNum] = kitAssemblyData[supplyKitTrackingNum];
+                        customParticipantObj[returnKitTrackingNum] = kitAssemblyData[returnKitTrackingNum];
+                        customParticipantObj[returnKitId] = kitAssemblyData[returnKitId];
+                        customParticipantObj[collectionCardId] = kitAssemblyData[collectionCardId];
+                    }
+                } catch (error) {
+                    console.error("Error in shippedKitStatusParticipants, fetching kit assembly document.", error);
+                    throw new Error("Error in shippedKitStatusParticipants, fetching kit assembly document.", error);
+                }
+                shippedStatusParticipants.push(customParticipantObj)
+            }
+        }
+        
+        return shippedStatusParticipants;
```

Comment:
I was thinking about how else to speed this up (especially since we know the data structure needs to be optimized in the future) and I believe you can use `Promise.all()` here for the second query instead of awaiting them individually.

Please test & confirm if you use this. Thanks

```javascript
if (snapshot.size > 0) {
    const participants = [];
    const kitAssemblyPromises = [];

    for (const doc of snapshot.docs) {
        const data = doc.data();
        const participantConnectID = data['Connect_ID'];
        participants.push({
            "Connect_ID": participantConnectID,
            [healthCareProvider]: data[healthCareProvider],
            [shippedDateTime]: data[collectionDetails]?.[baseline]?.[bioKitMouthwash]?.[shippedDateTime] || '',
            [mouthwashSurveyCompletionStatus]: data[mouthwashSurveyCompletionStatus],
        });

        kitAssemblyPromises.push(
            db.collection("kitAssembly")
                .where('Connect_ID', '==', participantConnectID)
                .select(`${supplyKitId}`, `${supplyKitTrackingNum}`, `${returnKitTrackingNum}`, `${returnKitId}`, `${collectionCardId}`)
                .get()
        );
    }

    // await the Promise.all. These will be ordered the same as the participant docs since they're going into Promise.all in the same order.
    const kitAssemblySnapshot = await Promise.all(kitAssemblyPromises);
    kitAssemblySnapshot.forEach((result, index) => {
        // If the kit assembly document exists, assign the data to the participant object.
        // If it doesn't exist, the result will have an empty placeholder. Skip if empty.
        if (!result.empty) {
            const kitData = result.docs[0].data();
            // In-place assignment of the kit data to the custom participant object
            Object.assign(participants[index], kitData);
        }
    });

    return participants;
```

Reply:
Thanks for the optimization suggestion, I'll test and try this out.